File: YEAR 2/SEM 1/Advanced Programming Methods/LAB 4/src/Util/Dictionary/MyDictionary.java
package Util.Dictionary;

import java.util.*;

public class MyDictionary<K,V> implements MyIDictionary<K,V> {
    private Map<K,V> dictionary;

    public MyDictionary() {
        dictionary = new HashMap<>();
    }

    @Override
    public void put(K key, V value) {
        dictionary.put(key,value);
    }

    @Override
    public V get(Object key) {
        return dictionary.get(key);
    }

    @Override
    public boolean containsKey(Object key) {
        return dictionary.containsKey(key);
    }

    @Override
    public boolean containsValue(Object value) {
        return dictionary.containsValue(value);
    }

    @Override
    public boolean isEmpty() {
        return dictionary.isEmpty();
    }

    @Override
    public V remove(K key) {
        return dictionary.remove(key);
    }

    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder();

        for (Object o : this.dictionary.entrySet()) {
            Map.Entry pair = (Map.Entry) o;
            stringBuilder.append("{").append(pair.getKey()).append(";").append(pair.getValue()).append("}");
            stringBuilder.append("\r\n");
        }

        return stringBuilder.toString();
    }

    @Override
    public int size() {
        return this.dictionary.size();
    }

    @Override
    public Collection<V> values() {
        return this.dictionary.values();
    }
}

File: YEAR 2/SEM 1/Advanced Programming Methods/ToyLanguage/ToyLanguage/Util/Dictionary/MyDictionary.cs
using System.Collections.Generic;
using System.Text;


namespace ToyLanguage.Util.Dictionary
{
    class MyDictionary<TKey, TValue> : MyIDictionary<TKey, TValue>
    {
        
        private readonly Dictionary<TKey, TValue> dictionary = new Dictionary<TKey, TValue>();

        public int Count { get => dictionary.Count; }

        public void Add(TKey key, TValue value)
        {
            if (key == null) throw new MyDictionaryException("Key cannot be null!");

            if (this.dictionary.ContainsKey(key)) throw new MyDictionaryException("Duplicate key!");

            dictionary.Add(key, value);
        }

        public bool ContainsKey(TKey key)
        {
            return this.dictionary.ContainsKey(key);
        }

        public bool Remove(TKey key)
        {
            if (key == null) throw new MyDictionaryException("Key cannot be null!");

            return dictionary.Remove(key);
        }

        public bool TryGetValue(TKey key, out TValue value)
        {
            if(key == null) throw new MyDictionaryException("Key cannot be null!");

            return dictionary.TryGetValue(key, out value);
        }

        public void Update(TKey key, TValue value)
        {
            if(key == null) throw new MyDictionaryException("Key cannot be null!");
            if(!dictionary.ContainsKey(key)) throw new MyDictionaryException("Key is not in the dictionary!");

            dictionary.Remove(key);
            dictionary.Add(key, value);
        }

        public Dictionary<TKey,TValue>.Enumerator GetEnumerator()
        {
            return dictionary.GetEnumerator();
        }

        public override string ToString()
        {
            StringBuilder builder = new StringBuilder();

            foreach(var item in dictionary)
            {
                builder.Append("{" + item.Key + ";" + item.Value + ")");
                builder.Append(System.Environment.NewLine);
            }

            return builder.ToString();
        }
    }
}
